composeApp/src/commonMain/kotlin/sp/bvantur/tasky/home/data/HomeRepositoryImpl.kt

```diff
@@ -57,4 +59,19 @@ class HomeRepositoryImpl(
     override suspend fun logoutUser(): TaskyEmptyResult<TaskyError> = localDataSource.clearLocalData().onSuccess {
         remoteDataSource.logoutUser()
     }
+
+    override suspend fun deleteEventById(id: String): TaskyEmptyResult<TaskyError> {
+        val localResult = localDataSource.deleteEventById(id)
+
+        if (localResult.isError()) {
+            // TODO handle database error
+            return localResult.asEmptyDataResult()
+        }
+        return eventRemoteDataSource.deleteEventById(id).onError {
+            localDataSource.changeSyncStepForEvent(
+                localResult.getSuccessResultOrNull() ?: return@onError,
+                SyncStep.DELETE
+            )
```

Comment:
Don’t you think this would be a good place to schedule a worker to perform a sync?

Reply:
Currently, another flow will pick up SyncStep.DELETE from the database and schedule a new worker to remove an event. Do you see any problems with my current approach?

Reply:
The logic here is split for no apparent reason to me. As a new developer, I wouldn’t know that a worker needs to be scheduled when an event gets deleted because I won’t know that the table is being reacted to in a different place in the codebase. The sync worker getting scheduled is directly tied to the local item being deleted. I’d keep that logic as close as possible - especially because by doing it here, I can protect myself from the coroutine getting cancelled when just after the agenda item is cancelled locally by calling scheduling the sync worker inside a coroutine launched by an application-scope coroutine.

Reply:
Ok, if I understand correctly you would leave the syncing flow where I am listening for changes in Home screen as it is, and you would add a call to worker here anyway so it is more tied together with the current logic that is happening here.

Reply:
I don’t see where you are listening for these changes, but I suggest you _replace_ it with my suggestion.

Reply:
@Kai-Cilliers I am scheduling syncing of events this in HomeViewModel class. I can do it here as you are suggesting, but what will happen if syncing won't complete in the current session of the app and you need to retry syncing of this user action? How do you solve this then? I guess there needs to be a backup for syncing the data for such cases as I currently have in my HomeViewModel with `repository.syncPendingAgendaItems()` call.

Reply:
By _”...what wil happen if syncing won’t complete…”_, I can interpret it in two ways. I will cover both scenarios here with their solutions:

1. **The call to schedule the sync was cancelled.** This can happen when the user leaves the screen and the ViewModel is cleared and all coroutines launched from the `viewModelScope` will be cancelled. You can have a situation where you modify your database and the scope getting cancelled before you can move to the next line to make an API call or schedule a worker. To work around this, we suggest you make use of a `CoroutineScope` scoped to the application’s lifecycle.

```
class MyApp: Application() {

    val applicationScope = CoroutineScope(SupervisorJob())
    
    // ...
}
```
You can then provide an instance of this scope in your Koin module:
```
single<CoroutineScope> {
  (androidApplication() as MyApp).applicationScope
}
```
Then, inject it into the class you need it in:
```
class MyRepository(
    private val applicationScope: CoroutineScope,
    ...
)
```
Usage:
```
applicationScope.launch {
    // schedule sync call
}.join()
```
This way, if the coroutine reached this point and wasn't cancelled, we make sure that it also won't be cancelled during this second call. The `join()` makes sure to still suspend the outer suspend function until the coroutine is finished.

2. **The worker failed to succeed.** This scenario is handled when you construct your `Worker` and inside its implementation. 

First you can set the backoff policy (when the worker should try again when it does not succeed):
```
val workRequest = OneTimeWorkRequestBuilder<DeleteReminderWorker>()
  .setBackoffCriteria(
    backoffPolicy = BackoffPolicy.EXPONENTIAL,
    backoffDelay = 2000L,
    timeUnit = TimeUnit.MILLISECONDS
  )
  /* other configurations go here */
  .build()
```

And inside your worker, return `Result. retry()` when you want the worker to schedule itself again to execute itself.
```
class DeleteReminderWorker(
  appContext: Context,
  private val params: WorkerParameters,
  private val remoteDataSource: RemoteAgendaDataSource,
  private val pendingSyncDao: PendingSyncDao,
) : CoroutineWorker(appContext, params) {

  override suspend fun doWork(): Result {
    if (runAttemptCount >= 5) {
      return Result.failure()
    }

    val pendingReminderId = params.inputData.getString(REMINDER_ID) ?: return Result.failure()
    val pendingEntity = pendingSyncDao.getDeletedReminderSyncEntity(pendingReminderId) ?: return Result.failure()

    return when (val result = remoteDataSource.deleteAgendaItem(pendingEntity.reminderId)) {
      is com.plcoding.tasky.core.domain.util.Result.Error -> {
        // toWorkerResult() is an extension function that maps the domain model DataError to either Result.failure() or Result.retry()
        result.error.toWorkerResult()
      }

      is com.plcoding.tasky.core.domain.util.Result.Success -> {
        pendingSyncDao.deleteDeletedReminderSyncEntity(pendingReminderId)
        Result.success()
      }
    }
  }
}
```

I hope this accurately answers your question 😊

Reply:
Ok, I got what you mean now. Still, I am not sure if this is the best approach for the KMP project. From what I read, WorkManager needs to be set only once and it will always complete its work, even if you restart your phone. Because of that, you can easily set syncing whenever the remote communication fails and you don't need to set it again in some special conditions. I am not sure if that is the case for iOS for example. Or maybe some other target that can be supported with KMP. What would your approach be then if you don't have such capability to set syncing one and execute it in every condition?

Reply:
I see what you mean 🤔 I would first go see what libraries there are for iOS and desktop platforms to handle such background processes. If they have, great! I’ll create an abstraction to and provide the specific actual implementation for each platform. If not, then first solution that comes to mind would be implement the background solution to android and whatever other platform supports background work scheduling, and on the other platforms, I’d determine places in the app to check for any pending syncs and run them.